File: ast/src/testing/vanila/php_minimal/controllers/UserController.php
<?php

require_once __DIR__ . '/../models/User.php';
require_once __DIR__ . '/../includes/helpers.php';

class UserController {
    public function index() {
        $users = User::all();
        jsonResponse($users);
    }

    public function show($id) {
        $user = User::find($id);
        if ($user) {
            jsonResponse($user);
        } else {
            jsonResponse(['error' => 'User not found'], 404);
        }
    }

    public function create() {
        $data = json_decode(file_get_contents('php://input'), true);
        if (!isset($data['username']) || !isset($data['email'])) {
            jsonResponse(['error' => 'Invalid input'], 400);
        }

        $user = new User(sanitize($data['username']), sanitize($data['email']));
        $user->save();
        
        jsonResponse(['id' => $user->id, 'message' => 'User created'], 201);
    }
}

File: ast/src/testing/vanila/ruby_minimal/app/controllers/user_controller.rb
require_relative '../../lib/models'
require_relative '../../lib/utils'

class UserController
  def self.index
    users = Models::User.all.map(&:to_h)
    Utils.json_response(users)
  end

  def self.show(id)
    user = Models::User.find(id)
    if user
      Utils.json_response(user.to_h)
    else
      Utils.json_response({ error: 'User not found' }, 404)
    end
  end

  def self.create(params)
    user = Models::User.new(params)
    if user.save
      Utils.json_response(user.to_h, 201)
    else
      Utils.json_response({ error: 'Invalid user data' }, 422)
    end
  end
end
